Alerting: Fetch configuration from the database and run a notification service

gotjosh opened the pull request:
**What this PR does / why we need it**:

With this PR, we can now fetch configuration from the database and apply it at runtime against the running notification instance.

**Special notes for your reviewer**:

This supersedes the work done by #32107

gotjosh committed:
Alerting: Fetch configuration from the database and run a notification
instance

Co-Authored-By: Ganesh Vernekar <15064823+codesome@users.noreply.github.com>
Signed-off-by: Josue Abreu <josue@grafana.com>

gotjosh commented on pkg/services/ngalert/database_mig.go:
```diff
@@ -107,18 +107,3 @@ func alertInstanceMigration(mg *migrator.Migrator) {
 	mg.AddMigration("add index in alert_instance table on def_org_id, def_uid and current_state columns", migrator.NewAddIndexMigration(alertInstance, alertInstance.Indices[0]))
 	mg.AddMigration("add index in alert_instance table on def_org_id, current_state columns", migrator.NewAddIndexMigration(alertInstance, alertInstance.Indices[1]))
 }
-
```
Don't be deceived, this is not gone simply moved to where I think it's the right place.

codesome commented on pkg/services/ngalert/notifier/config.go:
```diff
@@ -0,0 +1,77 @@
+	"gopkg.in/yaml.v2"
+)
+
+func PersistTemplates(cfg *api.PostableUserConfig, path string) ([]string, bool, error) {
```
I think we can delete the old files at this point else it would cause file leak. I wish we could pass string directly instead of files :|

codesome commented on pkg/services/ngalert/notifier/alertmanager.go:
```diff
@@ -7,9 +7,19 @@ import (
 	"sync"
 	"time"
 
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier/channels"
+
+	"github.com/prometheus/alertmanager/template"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+
+	"github.com/grafana/grafana/pkg/services/sqlstore"
```
We can group these imports better with the ones below

codesome commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/grafana/grafana/pkg/services/ngalert/notifier"
```
We can group it with the below imports

codesome commented on pkg/services/ngalert/notifier/alertmanager.go:
```diff
@@ -70,72 +87,110 @@ func (am *Alertmanager) Init() (err error) {
 	}
 }
 
+// AddMigration runs the database migrations as the service starts.
+func (am *Alertmanager) AddMigration(mg *migrator.Migrator) {
+	alertmanagerConfigurationMigration(mg)
+}
+
```
Am I missing something or is it not being not called anywhere now after you moved `alertmanagerConfigurationMigration`? Also we could just expose `alertmanagerConfigurationMigration` and not needed to go through Alertmanager.

gotjosh replied:
Not sure I get this statement. `AddMigration` is part of Grafana's service interface, it is run by the DI framework to call this method as we initilize the service:

```
// DatabaseMigrator allows the caller to add migrations to
// the migrator passed as argument
type DatabaseMigrator interface {
	// AddMigrations allows the service to add migrations to
	// the database migrator.
	AddMigration(mg *migrator.Migrator)
}
```

papagian replied:
> Not sure I get this statement. AddMigration is part of Grafana's service interface, it is run by the DI framework to call this method as we initilize the service:

exactly! check also [this](https://github.com/grafana/grafana/blob/master/contribute/architecture/backend/database.md#implement-databasemigrator)

codesome commented on pkg/services/ngalert/notifier/alertmanager.go:
```diff
@@ -70,72 +87,110 @@ func (am *Alertmanager) Init() (err error) {
 func (am *Alertmanager) Run(ctx context.Context) error {
 	// Make sure dispatcher starts. We can tolerate future reload failures.
-	if err := am.ReloadConfigFromDatabase(); err != nil {
+	if err := am.SyncAndApplyConfigFromDatabase(); err != nil && errors.Is(err, store.ErrNoAlertmanagerConfiguration) {
```
Did you mean `!errors.Is`?

gotjosh commented on pkg/services/ngalert/notifier/config.go:
```diff
@@ -0,0 +1,77 @@
+		}
+		// nolint:gosec
+
+		templatesChanged = true
```
Good catch.

gotjosh committed:
Address review comments

gotjosh committed:
More review comments

gotjosh commented on pkg/services/ngalert/notifier/config_test.go:
```diff
@@ -0,0 +1,122 @@
+		{
+			name:           "With valid templates file names, it persists successfully",
+			templates:      map[string]string{"email.template": "a perfectly fine template"},
+			expectedChange: true,
```
How does this set looks?

codesome replied:
Perfect!

gotjosh committed:
Appease the linter

papagian commented on pkg/services/ngalert/api/api.go:
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/grafana/alerting-api/pkg/api"
```
This conflicts with the package's name; I would suggest using an alias.

papagian commented on pkg/services/ngalert/api/api.go:
```diff
@@ -38,6 +44,7 @@ type API struct {
 	Schedule        schedule.ScheduleService
 	Store           store.Store
 	DataProxy       *datasourceproxy.DatasourceProxyService
+	Alertmanager    Alertmanager
```
Is it required? I don't think it's used.

gotjosh replied:
It is, otherwise the interface the interface is unused.

papagian replied:
what do you mean unused?

gotjosh replied:
The interface we defined above which is mean to serve the API struct - I mean, I can remove it but then we have no way to signal where this interface is meant to be used which can be a source of confusion.

codesome replied:
We will use it soon in future PRs, I guess?

gotjosh replied:
I was expecting @papagian to use in this alertmanager API implementation PR.

papagian replied:
ok, I see what you mean now! I was thinking of calling this in store package when storing the configuration but maybe I can move it in the API level...

papagian commented on pkg/services/ngalert/notifier/config_test.go:
```diff
@@ -0,0 +1,154 @@
+		},
+		{
+			name:          "With a invalid filename, it fails",
+			templates:     map[string]string{"adirecotry/email.template": "a perfectly fine template"},
+			expectedError: errors.New("template file name 'adirecotry/email.template' is  not valid"),
```
```suggestion
			templates:     map[string]string{"adirectory/email.template": "a perfectly fine template"},
			expectedError: errors.New("template file name 'adirectory/email.template' is not valid"),
```

papagian commented on pkg/services/ngalert/notifier/config.go:
```diff
@@ -0,0 +1,99 @@
+	pathSet := map[string]struct{}{}
+	for name, content := range cfg.TemplateFiles {
+		if name != filepath.Base(filepath.Clean(name)) {
+			return nil, false, fmt.Errorf("template file name '%s' is  not valid", name)
```
```suggestion
			return nil, false, fmt.Errorf("template file name '%s' is not valid", name)
```

papagian reviewed (approved):
AFAICT LGTM
I left some minor comments

gotjosh committed:
Apply suggestions from code review

Co-authored-by: Sofia Papagiannaki <papagian@users.noreply.github.com>

gotjosh committed:
Address more reviews

codesome commented on pkg/services/ngalert/notifier/config.go:
```diff
@@ -0,0 +1,99 @@
+		templatesChanged = true
+	}
+
+	// Now that we have the list of _actual_ templates, let's remove the ones that we don't need.
+	existingFiles, err := ioutil.ReadDir(path)
+	if err != nil {
+		log.Error("unable to read directory for deleting alertmanager templates", "err", err, "path", path)
+	}
+	for _, existingFile := range existingFiles {
+		p := filepath.Join(path, existingFile.Name())
+		_, ok := pathSet[p]
+		if !ok {
+			templatesChanged = true
+			err := os.Remove(p)
+			if err != nil {
+				log.Error("unable to delete template", "err", err, "file", p)
+			}
+		}
+	}
```
We are not using the returned `templatesChanged` boolean right now but I can see how this logic can cause issue: when we use it and skip template re-creation, then on restart with same disk, this logic is going to find that the files are same as config and say that templates didn't change, and hence we might skip creation of template tree. So the safest way to do it would be having the list of files loaded in the memory and compare against that, hence on restart we won't miss it.

codesome replied:
But I guess that is nothing to do with `PersistTemplates` itself, it is doing its job. We could take care in alertmanager to not skip templates when it's loading for the first time. So nothing to do for now, but something to be kept in mind.

gotjosh replied:
Discussed this offline, it's all about the usage of `PersistTemplates` return values. Decided to leave it as is for now and we are not yet using the variable.

In my eyes, this is meant to save persisting the file to disk over and over as we poll for new configuration.

gotjosh committed:
Address CI issues

- We do not have a default registry during Init
- Update the alerting-api depdency

gotjosh committed:
Remove the name on receivers

gotjosh committed:
One more
